chain/d3nlib/client.go

```diff
@@ -0,0 +1,67 @@
+package d3nlib
+
+import (
+	"sync"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/tendermint/tendermint/crypto"
+)
+
+type BandStatefulClient struct {
+	sequenceNumber uint64
+
+	mtx      sync.RWMutex
+	provider BandProvider
+}
+
+func NewBandStatefulClient(privKey crypto.PrivKey) (BandStatefulClient, error) {
+	provider, err := NewBandProvider(privKey)
+	if err != nil {
+		return BandStatefulClient{}, err
+	}
+	_, seq, err := authtypes.NewAccountRetriever(provider.cliCtx).GetAccountNumberSequence(provider.Sender())
+	if err != nil {
+		return BandStatefulClient{}, err
+	}
+
+	return BandStatefulClient{
+		sequenceNumber: seq,
+		provider:       provider,
+	}, nil
+}
+
+func (client *BandStatefulClient) SendTransaction(
+	msg sdk.Msg, gas uint64,
+	memo, fees, gasPrices, broadcastMode string,
+) (sdk.TxResponse, error) {
+	// Ask current sequence number
+	_, seq, err := authtypes.NewAccountRetriever(client.provider.cliCtx).
+		GetAccountNumberSequence(client.provider.Sender())
+	if err != nil {
+		return sdk.TxResponse{}, err
+	}
+	client.mtx.Lock()
```

Comment:
This is fine as is since it's obvious that nothing between lock and unlock can fail. But you can also write it like this.

```
// YOUR OTHER CODE
{
     client.mtx.Lock()
     defer client.mtx.Unlock()
     // YOUR CODE HERE UNDER LOCK
}
// YOUR OTHER CODE
```